### ISSUE TITLE
Long utf8_strings never seem to compile.

### ISSUE THREAD

**Issue by tpendragon** [plataformatec/nimble_parsec#21]
If I do the following:
```elixir
defmodule MarcParser do
  import NimbleParsec

  defparsec :parse_marc,
    utf8_string([], 24)
end
```

and run `mix compile`, it never finishes.

**Comment by DavidAntaramian** [comment]
I can confirm this using NimbleParsec 0.2.2 and 0.3.0 and Elixir 1.6.4 and Erlang/OTP 20.3.1.

After some non-scientific testing using `time mix compile`, it seems that the compilation time is growing exponentially in relation to the combinator's length parameter.

![chart](https://user-images.githubusercontent.com/156529/38578732-c6fec44a-3cd2-11e8-9c92-23086e39bdd3.png)

| String Length | Time v0.2.2 (seconds) | Time v0.3.0 (seconds) |
|---------------|----------------------:|----------------------:|
| 1             | 0.990                 | 0.971                 |
| 2             | 0.870                 | 0.909                 |
| 3             | 0.796                 | 0.960                 |
| 4             | 0.793                 | 0.964                 |
| 5             | 0.828                 | 1.147                 |
| 6             | 0.929                 | 1.015                 |
| 7             | 0.944                 | 1.065                 |
| 8             | 1.101                 | 1.502                 |
| 9             | 1.574                 | 1.409                 |
| 10            | 2.569                 | 2.111                 |
| 11            | 4.183                 | 3.782                 |
| 12            | 7.766                 | 7.229                 |
| 13            | 17.990                | 14.937                |
| 14            | 40.444                | 31.696                |
| 15            | 88.130                | 66.950                |

**Comment by tpendragon** [comment]
@josevalim Thanks for taking a look at this - can you say anything about why this was closed? If it's just something that shouldn't be done, maybe there's some pointers I can look over to see if I missed something?

**Comment by josevalim** [comment]
Fixed in 0.3.1.
